### DISCUSSION TITLE
Should I use SB or Driver for threading?

### DISCUSSION THREAD

**Discussion by misterpeople** [seleniumbase/SeleniumBase#2901] [Q&A] [answered]
So Im building a multi-step bot that needs to login interact with the website and scrape WITH multiple accounts.
Currently I use ThreadPoolExecutor to handle co-currency and threading

I use these methods below:
 with SB(uc=True, proxy=proxy_url, incognito=True) as sb:
sb.driver.uc_open_with_reconnect
 sb.wait_for_element_visible
  sb.open
  sb.driver.uc_click
  sb.click
 sb.select_option_by_index
 day = int(cell.find_element(by="css selector", value="a").text)
 row.find_element(by="css selector", value="input[type='radio']").click()
 
 sometimes I get detected so I guess IM not doing this stealthy enough, I am not sure if I should only use DRIVER methods or SB as well? 
 
 Also what is the differences between
   sb.click
   to
   sb.driver.uc_click
   
   aint both suppose to be stealthy?

**Comment by mdmintz** [comment, marked as answer]
There are some examples that fully get to a page and don't use Selenium at all to click:
* [SeleniumBase/examples/raw_brotector_captcha.py](https://github.com/seleniumbase/SeleniumBase/blob/master/examples/raw_brotector_captcha.py)
* [SeleniumBase/examples/raw_hobbit.py](https://github.com/seleniumbase/SeleniumBase/blob/master/examples/raw_hobbit.py)
* [SeleniumBase/examples/raw_antibot_login.py](https://github.com/seleniumbase/SeleniumBase/blob/master/examples/raw_antibot_login.py)

You might be able to learn from those.

`sb.click` is the same as `sb.driver.click`
`sb.uc_click` is the same as `sb.driver.uc_click`

Unless a `driver` method by the same name already existed with regular Selenium, the `sb.driver` version will be the same, although `sb` has more methods than `sb.driver`.

`uc_click()` uses JavaScript to do a stealthy click (more stealthy than `js_click()`), however CloudFlare detects JS clicks of any kind now, even if Selenium is not detected, and that's why `pyautogui` is needed now for actual stealthy clicking.

You'll need to use thread-locking to prevent overlaps with `pyautogui`. https://pypi.org/project/fasteners/ can help with that.

**Reply by misterpeople** [reply]
Thanks! my bot suppose to work simultaneously though with each account doing his own thing AND as I understand using pyautogui correctly will require me to STOP ALL other proccesses of all threads while it clicks on the button and then continue as normal with all the threads? 
Also consdering I will have to use pyautogui to click, EACH account will have to wait for the other to finish clicking the elements on the page, etc? it would be a whole slow mess no?

**Reply by mdmintz** [reply]
The thread-locking is unavoidable with `pyautogui`. A specific window needs to be active for the actions to work.

**Reply by misterpeople** [reply]
I see, BUT let's say I bypassed the CAPTCHA with a PyAutoGUI click or automatically. Can I continue to perform the necessary clicks on the page I scrape using `sb.click`, and if we encounter a Cloudflare check again, can we use `sb.uc_gui_handle_cf`? Or, if I use `sb.click` to click on elements I need and then randomly encounter a CAPTCHA, will it bypass it?

Additionally, does `sb.uc_gui_handle_cf` detect CAPTCHA automatically, or do I need to use an iframe like `sb.uc_gui_handle_cf("iframe[title*='challenge']")`?

**Reply by mdmintz** [reply]
You can use `sb.click` once you've bypassed the CAPTCHA. You only need thread-locking between setting the active window/element, and the `pyautogui` click. `sb.uc_gui_handle_cf()` will detect the CAPTCHA and be used only if needed. If the CAPTCHA is not in the first iframe, then you'll need to specify the selector when calling `sb.uc_gui_handle_cf()`
